src/kernel/chainstatemanager_opts.h

```diff
@@ -42,6 +44,7 @@ struct ChainstateManagerOpts {
     DBOptions block_tree_db{};
     DBOptions coins_db{};
     CoinsViewOptions coins_view{};
+    const std::shared_ptr<ValidationNotifications> notifications;
```

Comment:
Thank you for the time you took for these detailed comments. Fleshing this out to a more proper kernel interface sounds good to me. How about moving `ValidationNotifications` to the `KernelContext` instead? Or does that conflict with the vision you have for getting rid of `uiInterface` global?

Reply:
> Thank you for the time you took for these detailed comments. Fleshing this out to a more proper kernel interface sounds good to me. How about moving `ValidationNotifications` to the `KernelContext` instead? Or does that conflict with the vision you have for getting rid of `uiInterface` global?

Hmm, It doesn't make sense to me that an object which is supposed to receive notifications from the kernel would be owned by the kernel. Outside code that wants to receive notifications should be able to receive them and manage its own memory without transferring ownership.

Also yes the NotificationImpl can basically be a replacement for the uiInterface global, so it makes sense to put it in NodeContext or some place similar. It would be weird to put in the kernel and have to `static_cast` it when we can just store it directly in NodeContext.

Reply:
> Would suggest `ValidationNotifications& notifications` or `ValdiataionNotifications* notifcations{nullptr}` here,

I'd prefer the pointer type to do not bother about lifetime of variables that were used for assigning.

Reply:
> I'd prefer the pointer type to do not bother about lifetime of variables that were used for assigning.

I'd agree with you if we passing a string or some other passive data structure as an option to the kernel. Better to just use copy/move/reference counts to get the options struct to manage the memory rather than worry about lifetime of external variables.

But `kernel::Notifications` option is not just a passive set-and-forget option because it actively receives callbacks from the kernel. It will very likely own resources and contain pointers to other state. (Our `KernelNotifications` implementation will need to hold more state to get rid of `uiInterface` and `fWarned` globals.) I think it's important for lifetime of the object to be explicit and not use copies/movies/reference counts so other object lifetimes can be known and fixed, and we don't make memory management harder to debug in the long term just to avoid thinking about it in the short term.